akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -743,26 +765,30 @@ abstract class ShardCoordinator(
 
       case RebalanceDone(shard, ok) =>
         rebalanceWorkers -= sender()
-        if (ok)
-          log.debug("Rebalance shard [{}] completed successfully.", shard)
-        else
-          log.warning("Rebalance shard [{}] didn't complete within [{}].", shard, handOffTimeout.pretty)
-        // The shard could have been removed by ShardRegionTerminated
-        if (state.shards.contains(shard)) {
-          if (ok) {
+
+        if (ok) {
+          log.debug("Shard [{}] deallocation completed successfully.", shard)
+
+          // The shard could have been removed by ShardRegionTerminated
+          if (state.shards.contains(shard)) {
             update(ShardHomeDeallocated(shard)) { evt =>
-              log.debug("Shard [{}] deallocated after rebalance", shard)
+              log.debug("Shard [{}] deallocated after", shard)
               state = state.updated(evt)
               clearRebalanceInProgress(shard)
               allocateShardHomesForRememberEntities()
               self.tell(GetShardHome(shard), ignoreRef)
             }
           } else {
-            // rebalance not completed, graceful shutdown will be retried
-            gracefulShutdownInProgress -= state.shards(shard)
             clearRebalanceInProgress(shard)
           }
+
         } else {
+          log.warning("Shard [{}] deallocation didn't complete within [{}].", shard, handOffTimeout.pretty)
+
+          // was that due to a graceful region shutdown?
+          // if so, consider the region as still alive and let it retry to gracefully shutdown later
+          gracefulShutdownInProgress -= state.shards(shard)
```

Comment:
I don't get this, if the region did shutdown gracefully there will be nobody retrying later?

Also, no couldn't the region have terminated, so that shard no longer is in `state.shards` and causing this to throw `java.util.NoSuchElementException`?

Reply:
You are right about the `NoSuchElementException`. Needs a guard.

About the graceful shutdown, this was already like that (see previous code).

This code only executes if `RebalanceDone(_, ok = false)`. It will be `false` if we get a timeout when waiting for the `BeginHandOffAck`.

If the timeout happens because of a delay and it was a graceful shutdown, then the region will send, soon or later, another `GracefulShutdownReq`. This is fine in that case. 

If the timeout happens because the node crashed, removing it from the set of `gracefulShutdownInProgress` is not an issue. Soon or later, it will be removed from the list of registered shard regions. 

However, this bring us to another discussion. Something that I have being trying to wrap my head around a few times while changing that code. 

If a node ever sends a `GracefulShutdownReq`, we should not bring it back to the set of active shard regions (like we are doing here). We don't want to allocate there anymore. We do want it to clear its cache and send buffered messages as soon as possible. We want to include it on rebalance, but only to send it hand-off messages.

Reply:
I also find that curious.

Reply:
NoSuchElementException? trying to remove an element from a Set that doesn't exist in the Set is fine, it will be same Set

Reply:
NoSuchElement when trying to find it in the Map.

Reply:
@johanandren, I think I understand it now.

If it get stuck, it will send another request to the coordinator, but the coordinator will only let it go if not on `gracefulShutdownInProgress`. See the handling of `GracefulShutdownReq` below. 

We need that dance between the parts. 

I think that it will be more efficient if we don't put it back on the list of active regions. It said it's leaving, so it needs to go. 

The region can keep try to shutdown itself and the coordinator can stop to allocate stuff on it and when we don't need all that back and forth.

Reply:
Anyway, I'm not suggesting we should change it now, but we can review this in baby-steps. For me is very helpful to wrap my mind around it. From time to time to small improvements will be beneficial.